### PR TITLE
IDENTITY-4390 Adding capability to publish Oauth related events

### PR DESCRIPTION
**Pull request by pulasthi7**

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java [hunk 1]
```diff
@@ -504,12 +525,31 @@ public OAuthRevocationResponseDTO revokeAuthzForAppsByResoureOwner(
...
             revokeRespDTO.setErrorMsg("Invalid revocation request");
+
+            //passing a single element array with null element to make sure listeners are triggered at least once
+            triggerPostRevokeListeners(oauthListeners, revokeRequestDTO, revokeRespDTO, new AccessTokenDO[]{null});
             return revokeRespDTO;
         }
         return new OAuthRevocationResponseDTO();
     }
 
+    private void triggerPostRevokeListeners(List<OauthEventListener> oauthListeners,
```
Comment (Pushpalanka) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/OAuthAdminService.java]: We can go for a seperate method to trigger pre listeners too. Isn't it? That way code would be more consistent..

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/event/OauthEventListener.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import org.wso2.carbon.identity.oauth2.dto.OAuth2AuthorizeRespDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationRequestDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationResponseDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuthRevocationRequestDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuthRevocationResponseDTO;
+import org.wso2.carbon.identity.oauth2.model.AccessTokenDO;
+import org.wso2.carbon.identity.oauth2.model.RefreshTokenValidationDataDO;
+import org.wso2.carbon.identity.oauth2.token.OAuthTokenReqMessageContext;
+
+public interface OauthEventListener {
```
Comment (godwinamila) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/event/OauthEventListener.java]: Rename interface as OAuthEventListener

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 3]
```diff
@@ -233,6 +239,23 @@ public OAuthRevocationResponseDTO revokeTokenByOAuthClient(OAuthRevocationReques
         //fix here remove associated cache entry
         TokenMgtDAO tokenMgtDAO = new TokenMgtDAO();
         OAuthRevocationResponseDTO revokeResponseDTO = new OAuthRevocationResponseDTO();
+        List<OauthEventListener> oauthListeners = OAuthComponentServiceHolder.getInstance().getOauthEventListeners();
```
Comment (godwinamila) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Shall we add this in to a method ?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2TokenValidationService.java [hunk 4]
```diff
@@ -40,57 +44,80 @@
      */
     public OAuth2TokenValidationResponseDTO validate(OAuth2TokenValidationRequestDTO validationReqDTO) {
 
-	TokenValidationHandler validationHandler = TokenValidationHandler.getInstance();
+        TokenValidationHandler validationHandler = TokenValidationHandler.getInstance();
+        List<OauthEventListener> oauthEventListeners =
+                OAuthComponentServiceHolder.getInstance().getOauthEventListeners();
+        for (OauthEventListener oauthEventListener : oauthEventListeners) {
```
Comment (godwinamila) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2TokenValidationService.java]: Put this event triggering into a different method

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/TokenResponseTypeHandler.java [hunk 5]
```diff
@@ -50,6 +53,13 @@
     public OAuth2AuthorizeRespDTO issue(OAuthAuthzReqMessageContext oauthAuthzMsgCtx)
             throws IdentityOAuth2Exception {
 
+        List<OauthEventListener> oauthListeners = OAuthComponentServiceHolder.getInstance().getOauthEventListeners();
```
Comment (godwinamila) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/TokenResponseTypeHandler.java]: Add into a different method

**Pull request merged**